src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -425,6 +425,12 @@ internal void AdvanceReader(in SequencePosition consumed, in SequencePosition ex
 
         private void AdvanceReader(BufferSegment consumedSegment, int consumedIndex, BufferSegment examinedSegment, int examinedIndex)
         {
+            // Throw if examined < consumed
+            if (consumedSegment != null && consumedSegment != null && GetLength(consumedSegment, consumedIndex, examinedSegment, examinedIndex) < 0)
```

Comment:
```suggestion
            if (consumedSegment != null && examinedSegment != null && GetLength(consumedSegment, consumedIndex, examinedSegment, examinedIndex) < 0)
```

Reply:
Also does this cover the case where consumed is at the end of a segment and examined is at the beginning and visa virsa? I'd assume GetLength handles it.

Reply:
I’m not sure it does, the length would be 0 I think

Reply:
@halter73 came up with an interesting example yesterday.

Examined is at the end of the block and consumed at the beginning of the next one:

```
  block 0        block 1
[        ] -> [          ]
        ^      ^
      exam   cons
```

When calling Advance with these cursors it woud succeed because `lenght == 0` and lastExamined would point to block 0. At the same time the segment `lastExamined` was pointing to would be returned to the segment pool and possibly reused ending up with situation where lastExamined moved all the way forward:

```
  block 1        block 0
[          ] -> [          ]
         ^                ^
       cons              exam
```

Reply:
Yes I spoke to @halter73 as well.

Reply:
While I still think there's a problem with _lastExamined, this check seems fine though since neither consumed nor examined are returned to the pool.

Reply:
I had an idea on how to solve this using running index but I wont put it in this change.